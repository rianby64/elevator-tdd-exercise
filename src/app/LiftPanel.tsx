
import * as React from 'react';

interface LiftPanelProps {
    floors: number;
    currentLevel: number;
}

interface LiftPanelState {
    highlightButtons: Set<number>;
}

export class LiftPanel extends React.Component<LiftPanelProps> {

    public state: LiftPanelState = {
        highlightButtons: new Set([])
    }

    private highlightButton(i: number): () => void {
        return (): void => {
            this.setState((state: LiftPanelState): LiftPanelState => {
                const highlightButtons = new Set([...state.highlightButtons, i]);
                return {...state, highlightButtons: highlightButtons};
            });
        }
    }

    private createButtons = (): JSX.Element[] => new Array(this.props.floors)
        .fill(undefined)
        .map((v: undefined, i: number): JSX.Element => {
            return <li
                key={i}
                className={this.props.currentLevel === (i + 1) ?
                    "active" : undefined}>
                <button
                    className={this.state.highlightButtons.has(i + 1) ?
                        "highlight" : undefined}
                    onClick={this.highlightButton(i + 1)}>{i + 1}</button>
            </li>;
        })
        .reverse()

    public render(): JSX.Element {
        return (
            <div>
                <span>Lift&apos;s panel</span>
                <ul>
                    {this.createButtons()}
                </ul>
            </div>
        );
    }
}
