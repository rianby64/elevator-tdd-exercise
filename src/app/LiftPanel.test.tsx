
import * as React from 'react';
import { mount } from 'enzyme';
import { LiftPanel } from './LiftPanel';

test('LiftPanel highlighs the currentLevel', (): void => {
    const el = (<LiftPanel currentLevel={3} floors={5} />);
    const sh = mount(el);
    expect(sh).toMatchSnapshot();
});

test('LiftPanel highlighs the pressed button', (): void => {
    const el = (<LiftPanel currentLevel={3} floors={5} />);
    const sh = mount(el);
    sh.findWhere(el => el.key() === '2').find('button').simulate('click');
    expect(sh).toMatchSnapshot();
});
