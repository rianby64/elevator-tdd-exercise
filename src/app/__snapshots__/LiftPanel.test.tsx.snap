// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiftPanel highlighs the currentLevel 1`] = `
<LiftPanel
  currentLevel={3}
  floors={5}
>
  <div>
    <span>
      Lift's panel
    </span>
    <ul>
      <li
        key="4"
      >
        <button
          onClick={[Function]}
        >
          5
        </button>
      </li>
      <li
        key="3"
      >
        <button
          onClick={[Function]}
        >
          4
        </button>
      </li>
      <li
        className="active"
        key="2"
      >
        <button
          onClick={[Function]}
        >
          3
        </button>
      </li>
      <li
        key="1"
      >
        <button
          onClick={[Function]}
        >
          2
        </button>
      </li>
      <li
        key="0"
      >
        <button
          onClick={[Function]}
        >
          1
        </button>
      </li>
    </ul>
  </div>
</LiftPanel>
`;

exports[`LiftPanel highlighs the pressed button 1`] = `
<LiftPanel
  currentLevel={3}
  floors={5}
>
  <div>
    <span>
      Lift's panel
    </span>
    <ul>
      <li
        key="4"
      >
        <button
          onClick={[Function]}
        >
          5
        </button>
      </li>
      <li
        key="3"
      >
        <button
          onClick={[Function]}
        >
          4
        </button>
      </li>
      <li
        className="active"
        key="2"
      >
        <button
          className="highlight"
          onClick={[Function]}
        >
          3
        </button>
      </li>
      <li
        key="1"
      >
        <button
          onClick={[Function]}
        >
          2
        </button>
      </li>
      <li
        key="0"
      >
        <button
          onClick={[Function]}
        >
          1
        </button>
      </li>
    </ul>
  </div>
</LiftPanel>
`;
